src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubReceiver.java

```diff
@@ -0,0 +1,386 @@
+      }
+
+    }
+
+    if (isStopped()) {
+      return;
+    }
+
+    throw new RuntimeException("Failed to create subscription after 5 attempts.");
+  }
+
+  /**
+   * Receive new messages and store based on the Storage Level settings.
+   *
+   * @throws ApiException     when the Pub/Sub API throws a non-retryable exception.
+   * @throws RuntimeException when the GrpcSubscriberStub cannot be created.
+   */
+  public void receive() {
+    SubscriberStubSettings subscriberStubSettings = getSubscriberStubSettings();
+
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Receiver Started execution");
+    }
+
+    try (SubscriberStub subscriber = getSubscriberStub(subscriberStubSettings)) {
+      String subscriptionName = ProjectSubscriptionName.format(project, subscription);
+      fetchMessagesWithRetry(subscriber, subscriptionName);
+    } catch (IOException | ApiException e) {
+      throw new RuntimeException("Failed to fetch new messages.", e);
+    }
+
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Receiver completed execution");
+    }
+  }
+
+  /**
+   * Fetch new messages for our subscription.
+   * Implements exponential backoff strategy when a retryable exception is received.
+   *
+   * @param subscriber       The subscriber stub.
+   * @param subscriptionName The name of the subscription to use to pull data.
+   * @throws ApiException when the Pub/Sub API throws a non-retryable exception.
+   */
+  protected void fetchMessagesWithRetry(SubscriberStub subscriber, String subscriptionName) {
+    int backoff = backoffConfig.getInitialBackoffMs();
+
+    while (!isStopped()) {
+      try {
+        fetchAndAck(subscriber, subscriptionName);
+        backoff = backoffConfig.getInitialBackoffMs();
+      } catch (ApiException ae) {
+        if (ae.isRetryable()) {
+          backoff = sleepAndIncreaseBackoff(backoff);
+        } else {
+          throw ae;
+        }
+      }
+    }
+  }
+
+  /**
+   * Fetch new messages, store in Spark's memory, and ack messages.
+   * Based on SubscribeSyncExample.java in Google's PubSub examples.
+   *
+   * @param subscriber       The subscriber stub.
+   * @param subscriptionName The name of the subscription to use to pull data.
+   * @throws ApiException when the Pull request or ACK request fail.
+   */
+  protected void fetchAndAck(SubscriberStub subscriber, String subscriptionName) {
+    PullRequest pullRequest =
+      PullRequest.newBuilder()
+        .setMaxMessages(getMessageRate())
+        .setSubscription(subscriptionName)
+        .build();
+    PullResponse pullResponse = subscriber.pullCallable().call(pullRequest);
+
+    List<ReceivedMessage> receivedMessages = pullResponse.getReceivedMessagesList();
+    List<String> ackIds = receivedMessages.stream().map(ReceivedMessage::getAckId).collect(Collectors.toList());
+
+    //If there are no messages to process, continue.
+    if (receivedMessages.size() == 0) {
```

Comment:
This check should be before the `actIds` creation

Reply:
Good catch.

Reply:
Also, use `isEmpty()` instead of `size() == 0`